emissary-core/src/destination/session/mod.rs

```diff
@@ -3285,4 +3308,429 @@ mod tests {
+            inbound_private_key,
+            inbound_leaseset,
+        );
+
+        // create outbound `SessionManager`
+        let outbound_private_key = StaticPrivateKey::random(thread_rng());
+        let (
+            outbound_leaseset,
+            outbound_destination_id,
+            _outbound_destination,
+            _outbound_signing_key,
+        ) = {
+            let (leaseset, signing_key) = LeaseSet2::random();
+            let destination = leaseset.header.destination.clone();
+            let outbound_destination_id = leaseset.header.destination.id();
+
+            (
+                Bytes::from(leaseset.serialize(&signing_key)),
+                outbound_destination_id,
+                destination,
+                signing_key,
+            )
+        };
+        let mut outbound_session = SessionManager::<MockRuntime>::new(
+            outbound_destination_id.clone(),
+            outbound_private_key.clone(),
+            outbound_leaseset,
+        );
+        outbound_session.add_remote_destination(inbound_destination_id.clone(), inbound_public_key);
+
+        // send and handle NS message
+        let message = outbound_session.encrypt(&inbound_destination_id, vec![1u8; 4]).unwrap();
+        decrypt_and_verify!(&mut inbound_session, message, vec![1u8; 4]);
+
+        // send NSR as reply to the new NS
+        let message = inbound_session.encrypt(&outbound_destination_id, vec![2u8; 4]).unwrap();
+        decrypt_and_verify!(&mut outbound_session, message, vec![2u8; 4]);
+
+        // send ES message to bob, finalizing the session
+        let message = outbound_session.encrypt(&inbound_destination_id, vec![3u8; 4]).unwrap();
+        decrypt_and_verify!(&mut inbound_session, message, vec![3u8; 4]);
+
+        // send ES message to alice
+        let message = inbound_session.encrypt(&outbound_destination_id, vec![5u8; 4]).unwrap();
+        decrypt_and_verify!(&mut outbound_session, message, vec![5u8; 4]);
+
+        // Assert there is no pending next key
+        assert!(!inbound_session
+            .active
+            .get(&outbound_destination_id)
+            .expect("to exist")
+            .session
+            .has_pending_next_key());
+
+        // Exhaust tag set to force a next key
+        for _ in 0..SESSION_DH_RATCHET_THRESHOLD {
+            let message = outbound_session.encrypt(&inbound_destination_id, vec![7u8; 4]).unwrap();
+            decrypt_and_verify!(&mut inbound_session, message, vec![7u8; 4]);
+        }
+
+        // Assert we have a pending next key
+        assert!(inbound_session
+            .active
+            .get(&outbound_destination_id)
+            .expect("to exist")
+            .session
+            .has_pending_next_key());
+
+        // assert that nothing is ready immediately
+        assert!(
+            tokio::time::timeout(Duration::from_millis(100), inbound_session.next())
+                .await
+                .err()
+                .is_some()
+        );
```

Comment:
The 100ms timeout is somewhat random, I think this could be something along the lines of `assert!(inbound_session.next().now_or_never().is_none());`